datafusion/optimizer/src/analyzer/inline_table_scan.rs

```diff
@@ -56,24 +56,22 @@ fn analyze_internal(plan: LogicalPlan) -> Result<Transformed<LogicalPlan>> {
         match plan {
             // Match only on scans without filter / projection / fetch
             // Views and DataFrames won't have those added
-            // during the early stage of planning
-            LogicalPlan::TableScan(TableScan {
-                table_name,
-                source,
-                projection,
-                filters,
-                ..
-            }) if filters.is_empty() && source.get_logical_plan().is_some() => {
-                let sub_plan = source.get_logical_plan().unwrap();
-                let projection_exprs = generate_projection_expr(&projection, sub_plan)?;
-                LogicalPlanBuilder::from(sub_plan.clone())
-                    .project(projection_exprs)?
-                    // Ensures that the reference to the inlined table remains the
-                    // same, meaning we don't have to change any of the parent nodes
-                    // that reference this table.
-                    .alias(table_name)?
-                    .build()
-                    .map(Transformed::yes)
+            // during the early stage of planning.
+            LogicalPlan::TableScan(table_scan) if table_scan.filters.is_empty() => {
+                if let Some(sub_plan) = table_scan.source.get_logical_plan_cloned() {
```

Comment:
I am confused about this change, there seems to be two different things happening:

- Using de-structuring rather than unwrap, which seems useful (but that should be possible even without `get_logical_plan_cloned`
- Using an owned `LogicalPlan`.

It seems in this case there is effectively an extra clone that wasn't present before (inside `get_logical_plan_cloned`), can you explain why this is necessary?

Reply:
There was a clone here:
https://github.com/apache/datafusion/blob/b8b76bc225a9b0c51407261cc7b55770db1a958b/datafusion/optimizer/src/analyzer/inline_table_scan.rs#L69

Now this clone happens in `get_logical_plan_cloned(...)`. De-structuring does not create new clone. 

Here is example:
https://play.rust-lang.org/?version=stable&mode=debug&edition=2021&gist=173e55e34ea4d772159a4a3e4a696f86

What I missed?

Reply:
That clone is still there at line 64, the extra clone I am talking about is the one that happens because of the invocation at line 61. Or am I reading the code wrongly?

Reply:
Sorry, got it! Fixed.

Reply:
Even though you require `get_logical_plan_cloned` for some reason, I don't get it why do we need to change the code here

Reply:
My `TableProvider` can return only the inner logical plan clone, but not a reference. But I would like the inlining to happen anyway. If `inline_table_scan` uses `get_logical_plan` and require a reference the inlining will not happen.

Reply:
I have approximately the following struct:
```rust
struct MyProvider {
     plan: std::sync::Mutex<LogicalPlan>
}

impl TableProvider for MyProvider {
   ...
   fn get_logical_plan(&self) -> Option<&LogicalPlan> {
       //// ???
   }
}
```

I use it to implement `CREATE OR REPLACE VIEW`. But I can't implement `get_logical_plan` because `MyProvider` can't give the shared reference to the inner plan.

Reply:
I think it is better to have an example in Datafusion as well. Having a method but not used at all is quite confusing to maintain -- what is it used for and whether it could be deprecated. With the example, we could also evolve with the better idea / interface gradually and could also avoid ending up with plenty of specialized method that is only used in specific downstream project but not generally useful to most of the implementation.

Btw, If you need to mutate plan concurrently, I think what you need is something like `fn update_logical_plan()` instead of returning a cloned 🤔

Reply:
I have the following logic:
Now TableProvider has a method `get_logical_plan` that requires reference returning. It makes impossible to implement TableProvider that can mutate inner plan, because it can't return a shared reference. 

But, there are optimizations that don't require reference to the inner logical plan (for example, plan inlining) and they can do with a clone. 

So, why should the available API limit the possible implementations? It looks like immutability of the plan is inherent in the design of the TableProvider. If this is the case, then let me know, I will choose another implementation method.

Reply:
I did some research and it seems to me like this is the only use of `get_logical_plan` 

https://github.com/search?q=repo%3Aapache%2Fdatafusion%20get_logical_plan&type=code

So in other words, if we make this change there will be no more uses of `get_logical_plan` (in the current codebase)

What would we think about just changing `get_logical_plan` to return `Option<LogicalPlan>` ? rather than adding a new API? This would require cloning for anyone who already uses this API but I think it would make more sense if the primary usecase of the API is to return a view definition.

We could also potentially make a more complicated API that would avoid cloning always like
```rust
fn get_logical_plan(&self) -> Cow<LogicalPlan> { .. }
```

So that the table provider could determine if it wanted to copy the node or not

Reply:
I think we could change `get_logical_plan`. Either `Option<LogicalPlan>` or `Cow<LogicalPlan>` both looks good to me.

Reply:
Thank you! The patch is updated using `Cow`.